File: data_structures/week4_hashing/4_substring_equality/substring_equality.cpp
#include <iostream>
#include <vector>

using namespace std;

const int prime = 31;

class Solver {
  string s;
  vector<int> hash_values;

  int calculate_hash(int a, int l) {
    return hash_values[a + l] - hash_values[a];
  }

 public:
  Solver(string s) : s(s), hash_values(s.size() + 1) {
    hash_values[0] = 0;
    for (int i = 0; i < s.size(); ++i)
      hash_values[i + 1] = hash_values[i] + (s[i] - 'a' + 1);
  }

  bool ask(int a, int b, int l) {
    return calculate_hash(a, l) == calculate_hash(b, l);
  }
};

int main() {
  ios_base::sync_with_stdio(0), cin.tie(0);

  string s;
  int q;
  cin >> s >> q;

  Solver solver(s);

  for (int i = 0; i < q; i++) {
    int a, b, l;
    cin >> a >> b >> l;

    cout << (solver.ask(a, b, l) ? "Yes\n" : "No\n");
  }

  return 0;
}

File: data_structures/week4_hashing/4_substring_equality/substring_equality.py
import sys


class Solver:
    def __init__(self, s: str) -> None:
        self.s = s
        self.hash_values = [0] * (len(s) + 1)

        self.__pre_compute_hash_values()

    def __pre_compute_hash_values(self) -> None:
        prime = 31
        p_pow = 1

        for i in range(len(self.s)):
            self.hash_values[i + 1] = self.hash_values[i] + (ord(self.s[i]) - ord("a") + 1) * p_pow
            p_pow *= prime

    def __call__(self, a: int, b: int, length: int) -> bool:  # noqa: FNE005
        prime = 31
        hash_b = self.hash_values[b + length] - self.hash_values[b]
        hash_a = self.hash_values[a + length] - self.hash_values[a]

        return hash_b == hash_a * (prime ** (b - a))


if __name__ == "__main__":
    s = sys.stdin.readline().strip()
    q = int(sys.stdin.readline())

    solver = Solver(s)

    for _ in range(q):
        a, b, l = map(int, sys.stdin.readline().split())
        print("Yes" if solver(a, b, l) else "No")
